Source/CarthageKit/Xcode.swift

```diff
@@ -1080,85 +1080,152 @@ public typealias BuildSchemeProducer = SignalProducer<TaskEvent<(ProjectLocator,
+			// absolute link back to the project's Build folder.
 			do {
-				try NSFileManager.defaultManager().removeItemAtURL(dependencyBinariesURL)
-			} catch {
-				let dependencyParentURL = dependencyBinariesURL.URLByDeletingLastPathComponent!
-
-				do {
-					try NSFileManager.defaultManager().createDirectoryAtURL(dependencyParentURL, withIntermediateDirectories: true, attributes: nil)
-				} catch let error as NSError {
-					return .Failure(.WriteFailed(dependencyParentURL, error))
-				}
+				try NSFileManager.defaultManager().createSymbolicLinkAtURL(dependencyBinariesURL, withDestinationURL: rootBinariesURL)
+			} catch let error as NSError {
+				return .Failure(.WriteFailed(dependencyBinariesURL, error))
 			}
-
-			var isSymlink: AnyObject?
+		} else {
+			let linkDestinationPath = relativeLinkDestinationForDependency(dependency, subdirectory: CarthageBinariesFolderPath)
 			do {
-				try rawDependencyURL.getResourceValue(&isSymlink, forKey: NSURLIsSymbolicLinkKey)
+				try NSFileManager.defaultManager().createSymbolicLinkAtPath(dependencyBinariesURL.path!, withDestinationPath: linkDestinationPath)
 			} catch let error as NSError {
-				return .Failure(.ReadFailed(rawDependencyURL, error))
+				return .Failure(.WriteFailed(dependencyBinariesURL, error))
 			}
+		}
+		return .Success(true)
+	}
+}
+
+/// Creates symlink between the dependency checkouts and the root checkouts
+///
+/// Returns a signal indicating success
+private func symlinkCheckoutPathForDependencyProject(dependency: ProjectIdentifier, rootDirectoryURL: NSURL) -> SignalProducer<Bool, CarthageError> {
+	let rootCheckoutsURL = rootDirectoryURL.URLByAppendingPathComponent(CarthageProjectCheckoutsPath, isDirectory: true).URLByResolvingSymlinksInPath!
+	let rawDependencyURL = rootDirectoryURL.URLByAppendingPathComponent(dependency.relativePath, isDirectory: true)
+	let dependencyURL = rawDependencyURL.URLByResolvingSymlinksInPath!
+	let dependencyCheckoutsURL = dependencyURL.URLByAppendingPathComponent(CarthageProjectCheckoutsPath, isDirectory: true).URLByResolvingSymlinksInPath!
 
-			if isSymlink as? Bool == true {
-				// Since this dependency is itself a symlink, we'll create an
-				// absolute link back to the project's Build folder.
+	return SignalProducer<Bool, CarthageError>.attempt {
+		do {
+			try NSFileManager.defaultManager().createDirectoryAtURL(dependencyCheckoutsURL, withIntermediateDirectories: true, attributes: nil)
+		} catch let error as NSError {
+			if !(error.domain == NSCocoaErrorDomain && error.code == NSFileWriteFileExistsError) {
+				return .Failure(.WriteFailed(dependencyCheckoutsURL, error))
+			}
+		}
+		return .Success(true)
+	}
+	.flatMap(.Merge) { _ -> SignalProducer<Bool, CarthageError> in
+
+		return NSFileManager.defaultManager().carthage_enumeratorAtURL(rootCheckoutsURL, includingPropertiesForKeys: [NSURLIsDirectoryKey], options: [.SkipsSubdirectoryDescendants, .SkipsHiddenFiles])
```

Comment:
I think it'd be better to only create symlinks for the dependencies of this project.

So given these `Cartfile`s:

```
// App Cartfile
github "my/A"
github "my/B"
```

```
// my/A Cartfile
github "my/B"
```

```
// my/B Cartfile
```

You'd end up with this directory structure:

```
Carthage/
  Checkouts/
    A/
      Carthage/
        Checkouts/
          B -> ../../../B
    B/
       Carthage/
         Checkouts/
```